src/relay/ir/text_printer.cc

```diff
@@ -0,0 +1,752 @@
+    // TODO(marisha): add escape.
+    os << "\"" << value << "\"";
+  }
+  /*!
+   * \brief get a unique name with the corresponding prefix
+   * \param prefix The prefix of the name
+   * \return The returned name.
+   */
+  std::string GetUniqueName(std::string prefix) {
+    auto it = name_alloc_map_.find(prefix);
+    if (it != name_alloc_map_.end()) {
+      while (true) {
+        std::ostringstream os;
+        os << prefix << (++it->second);
+        std::string name = os.str();
+        if (name_alloc_map_.count(name) == 0) {
+          prefix = name;
+          break;
+        }
+      }
+    }
+    name_alloc_map_[prefix] = 0;
+    return prefix;
+  }
+  /*!
+   * \brief mark the beginning of a new scope
+   * \return The scope id.
+   */
+  int BeginScope() {
+    int sid = static_cast<int>(scope_valid_.size());
+    scope_valid_.push_back(true);
+    indent_ += 2;
+    return sid;
+  }
+  /*!
+   * \brief mark the end of an old scope.
+   * \param scope_id The scope id to be ended.
+   */
+  void EndScope(int scope_id) {
+    scope_valid_[scope_id] = false;
+    indent_ -= 2;
+  }
+  /*!
+   * \brief Print the indent to the stream.
+   */
+  void PrintIndent(int64_t more_indent = 0) {
+    for (int i = 0; i < indent_ + more_indent; ++i) {
+      stream_ << ' ';
+    }
+  }
+  /*!
+   * \brief print end of the line.
+   */
+  void PrintEndInst(const char* suffix) {
+    stream_ << suffix;
+  }
+  /*!
+   * \brief Allocate temporary value
+   * \return A new text value.
+   */
+  TextValue AllocTempVar() {
+    std::ostringstream os;
+    os << '%' << temp_var_counter_++;
+    return TextValue(os.str());
+  }
+  /*!
+   * \brief Allocate name to a variable.
+   * \param var The input variable.
+   * \return The corresponding name.
+   */
+  TextValue AllocVarName(const Var& var) {
+    std::string name = GetUniqueName('%' + var->name_hint);
+    TextValue val(name);
+    CHECK(!memo_.count(var));
+    memo_[var] = val;
+    return val;
+  }
+
+ private:
+  class AttrPrinter;
+  friend class AttrPrinter;
+  /*! \brief meta data context */
+  TextMetaDataContext meta_;
+  /*! \brief Check whether scope is still valid */
+  std::vector<bool> scope_valid_;
+  /*! \brief The current indentation value */
+  int indent_{0};
+  /*! \brief name allocation map */
+  std::unordered_map<std::string, int> name_alloc_map_;
+  /*! \brief Map from expression to its text value */
+  std::unordered_map<Expr, TextValue, NodeHash, NodeEqual> memo_;
+  /*! \brief counter of temporary variable */
+  int64_t temp_var_counter_{0};
+  /*! \brief Output stream */
+  std::ostringstream stream_;
+};
+
+/*!
+ * \brief Attribute printer which prints the attributes in the call.
+ */
+class TextPrinter::AttrPrinter: public AttrVisitor {
+ public:
+  AttrPrinter(std::ostream& stream, TextPrinter* parent)  // NOLINT(*)
+      : stream_(stream), parent_(parent) {}
+
+  void Visit(const char* key, double* value) final {
```

Comment:
Should the second parameter `value` in all the cases be `const type*`? Seems it won't be changed?

Reply:
This is a restriction of the AttrVisitor signature, as the same visitor is used to run setter workload as well